clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -232,6 +232,15 @@ public RuntimeException lastError() {
         return lastError;
     }
 
+    public synchronized boolean ensurePartitionAdded(TopicPartition tp) {
+        if (isInTransaction() && !partitionsInTransaction.contains(tp)) {
+            transitionToFatalError(new IllegalStateException("Attempted to dequeue a record batch to send " +
```

Comment:
It should not happen normally, but if there is a bug in the code, then the result is basically a corrupted topic, so I felt it is worth having the check. Transitioning to the fatal error state ensures that the user will see the error and that no further progress can be made. If we just threw the exception, the Sender would simply die apparently.